### PR TITLE
Move CRMI operations to Visitors

### PR DESCRIPTION
**Pull request by TahaAttari**

### CODE REVIEW

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/dstu3/ValueSetAdapter.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        return new Period();
+    }
+
+    public List<? extends ICompositeType> getRelatedArtifact() {
+        return new ArrayList<>();
+    }
+
+    public List<RelatedArtifact> getComponents() {
+        return new ArrayList<>();
+    }
```
Comment (TahaAttari) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/dstu3/ValueSetAdapter.java:126]: should getComponents return compose.include and expansion.contains for ValueSets?

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/r4/LibraryAdapter.java [hunk 2]
```diff
@@ -77,19 +90,65 @@ public boolean hasContent() {
...
 
     @Override
-    public List<ICompositeType> getContent() {
+    public List<Attachment> getContent() {
         return this.getLibrary().getContent().stream().collect(Collectors.toList());
     }
 
     @Override
-    public void setContent(List<ICompositeType> attachments) {
+    public <T extends ICompositeType> void setContent(List<T> attachments) {
```
Comment (TahaAttari) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/r4/LibraryAdapter.java:98]: couldn't figure out how to set this up properly due to Type Erasure

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/cql/evaluator/fhir/util/IDependencyInfo.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package org.opencds.cqf.cql.evaluator.fhir.util;
+
+import org.hl7.fhir.instance.model.api.IPrimitiveType;
+
+public interface IDependencyInfo {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/cql/evaluator/fhir/util/IDependencyInfo.java:5]: This interface is not used.  Is there a reason for that?

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/cql/evaluator/fhir/util/DependencyInfo.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package org.opencds.cqf.cql.evaluator.fhir.util;
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/cql/evaluator/fhir/util/DependencyInfo.java:1]: This needs to be put in the right place and the cql.evaluator folder removed.

File: cqf-fhir-utility/pom.xml [hunk 5]
```diff
@@ -21,6 +21,12 @@
       <artifactId>cqf-fhir-api</artifactId>
       <version>3.1.0-SNAPSHOT</version>
     </dependency>
+    <dependency>
+      <groupId>org.opencds.cqf.fhir</groupId>
+      <artifactId>cqf-fhir-test</artifactId>
+      <version>3.0.0</version>
```
Comment (barhodes) [inline cqf-fhir-utility/pom.xml:27]: Need to make sure this matches the current version in master.

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/SearchHelper.java [hunk 6]
```diff
@@ -84,6 +85,44 @@ IBaseResource searchRepositoryByCanonical(
 
         return result;
     }
+    /**
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/SearchHelper.java:88]: Nitpick: Line separation between methods

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java [hunk 7]
```diff
@@ -75,14 +79,13 @@ public <T extends IBaseResource, I extends IIdType> T read(
 
     @Override
     public <T extends IBaseResource> MethodOutcome create(T resource, Map<String, String> headers) {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java:81]: What is the reason for these changes?

Nitpick: use var when able

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java [hunk 8]
```diff
@@ -195,6 +198,28 @@ public <B extends IBaseBundle> B transaction(B transaction, Map<String, String>
         throw new NotImplementedException("The transaction operation is not currently supported");
     }
 
+    public static Bundle transactionStub(Bundle transaction, Repository repository) {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java:201]: This is R4 specific logic and cannot be in here.  This either needs to be done in a version agnostic way, (look to the BundleHelper) or with a factory that returns the version specific class based on the version of the Repository.

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/r4/PackageHelper.java [hunk 9]
```diff
@@ -159,11 +159,17 @@ public static BundleEntryComponent createEntry(Resource resource, boolean isPut)
...
+            var url = getUrl(resource);
+            if (hasVersion(resource)) {
+                entry.setFullUrl(url + "|" + getVersion(resource));
+                if (!isPut) {
                     request.setIfNoneExist(String.format("url=%s&version=%s", url, getVersion(resource)));
-                } else {
+                }
+            } else {
+                entry.setFullUrl(url);
+                if (!isPut) {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/r4/PackageHelper.java:172]: Please make sure these changes are also made in the Dstu3 and R5 versions of this class.

**Review by barhodes — CHANGES_REQUESTED** [review]
Some issues to address before this is ready to be merged.

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java (hunk 7 above)
Comment (TahaAttari) [reply cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java:80]: I needed to add `theId` of the created/updated resource to the `MethodOutcome` so that `transactionStub` could add it to the return bundle

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/IDependencyInfo.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    public String getReference();
+
+    public void setReference(String reference);
+
+    public String getReferencePackageId();
+
+    public void setReferencePackageId(String referencePackageId);
+
+    public List<? extends IBaseExtension> getExtension();
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/IDependencyInfo.java:19]:
```suggestion
    public List<? extends IBaseExtension<?, ?>> getExtension();
```

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/DependencyInfo.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+public class DependencyInfo implements IDependencyInfo {
+    // TODO: Need for figuring out how to determine which package the dependency is in.
+    private String referenceSource;
+    private String reference;
+    private String referencePackageId;
+    private List<? extends IBaseExtension> extensionList;
+
+    public DependencyInfo() {}
+
+    public DependencyInfo(String referenceSource, String reference, List<? extends IBaseExtension> extensionList) {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/DependencyInfo.java:16]:
```suggestion
    public DependencyInfo(String referenceSource, String reference, List<? extends IBaseExtension<?, ?>> extensionList) {
```

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/DependencyInfo.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        this.referenceSource = referenceSource;
+        this.reference = reference;
+        this.extensionList = extensionList;
+    }
+
+    public String getReferenceSource() {
+        return this.referenceSource;
+    }
+
+    public List<? extends IBaseExtension> getExtension() {
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/DependencyInfo.java:26]:
```suggestion
    public List<? extends IBaseExtension<?, ?>> getExtension() {
```

Comment (TahaAttari) [reply cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/DependencyInfo.java]: thanks 😁

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/KnowledgeArtifactAdapter.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+                        getDependencies().stream())
+                .collect(Collectors.toList());
+    }
+
+    <T extends ICompositeType & IBaseHasExtensions> void setRelatedArtifact(List<T> relatedArtifacts);
+
+    public IBase accept(KnowledgeArtifactVisitor visitor, Repository repository, IBaseParameters operationParameters);
+
+    @SuppressWarnings("unchecked")
+    default <T extends ICompositeType & IBaseHasExtensions> List<T> getOwnedRelatedArtifacts() {
```
Comment (TahaAttari) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/adapter/KnowledgeArtifactAdapter.java:93]: not sure why this is causing the Warning

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java [hunk 14]
```diff
@@ -95,9 +98,9 @@ public <I extends IIdType, P extends IBaseParameters> MethodOutcome patch(
 
     @Override
     public <T extends IBaseResource> MethodOutcome update(T resource, Map<String, String> headers) {
-        var outcome = new MethodOutcome();
         var resources = resourceMap.computeIfAbsent(resource.fhirType(), r -> new HashMap<>());
         var theId = resource.getIdElement().toUnqualifiedVersionless();
+        var outcome = new MethodOutcome(theId, false);
```
Comment (TahaAttari) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/repository/InMemoryFhirRepository.java:103]: outcome should have the ID of the newly updated resource

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/PackageHelper.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * This class consists exclusively of static methods that assist with packaging FHIR Resources.
+ */
+public class PackageHelper {
+
+    public static IBaseBackboneElement createEntry(IBaseResource resource, boolean isPut) {
+        final var fhirVersion = resource.getStructureFhirVersionEnum();
+        final var entry = BundleHelper.newEntryWithResource(resource.getStructureFhirVersionEnum(), resource);
+        String method;
+        var requestUrl = getResourceType(resource);
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/PackageHelper.java:19]:
```suggestion
        var requestUrl = resource.fhirType();
```

File: cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/PackageHelper.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+                        .toString();
+            case DSTU2:
+            case DSTU2_1:
+            case DSTU2_HL7ORG:
+            default:
+                throw new IllegalArgumentException(String.format(
+                        "Unsupported version of FHIR: %s",
+                        resource.getStructureFhirVersionEnum().getFhirVersionString()));
+        }
+    }
```
Comment (barhodes) [inline cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/PackageHelper.java:75]: IBaseResource already has a fhirType() method. This can be removed.

**Review by barhodes — APPROVED** [review]
